### PR TITLE
Adds request handler to update flaky tests

### PR DESCRIPTION
**Pull request by chunhtai**
Things that are still missing

1. https://github.com/flutterflakybot need to be given push access to the flutter/flutter
2. tests will be added once the design looks good
3. existing flaky issue need to be updated to have the predefined format for the scripts to pick up.

### CODE REVIEW

**Comment by chunhtai** [comment]
Hi @keyonghan can you take a look and see if the general design looks good to you, I will add the tests once it does.

File: app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+    final RepositoryContents contents = await client.repositories.getContents(config.flutterSlug, _testOwnerPath);
+    if (!contents.isFile) {
+      throw 'The path $_ciYamlPath should point to a file, but it is not!';
+    }
+    final String content = utf8.decode(base64.decode(contents.file.content.replaceAll('\n', '')));
+    return content;
+  }
+
+  Future<String> _findTestOwner(String testName, String testOwnersContent) async {
+    final List<String> lines = testOwnersContent.split('\n');
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart:381]: all `\n` has been replaced with ` `?

File: app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+        succeededBuilds: succeededBuilds ?? const <String>[],
+        recentCommit: row.f[5].v as String,
+        failedBuildOfRecentCommit: row.f[6].v as String,
+        testOwner: await _findTestOwner(testName, testOwnerContent)
+      );
+    }
+    return nameToStats;
+  }
+
+  Future<Map<String, ExistingGithubIssue>> _getExistingGithubIssues(GitHub client) async {
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart:138]: shall we inject repo to make this function general? We may extend to other repos in the future.

File: app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+    final Map<String, ExistingGithubIssue> nameToExistingIssue = <String, ExistingGithubIssue>{};
+    await for (final Issue issue in client.issues.listByRepo(
+        config.flutterSlug, state:'all', labels: <String>[kFlakeLabel])) {
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart]: shall we use the `severe: flake` label to limit the search range? `team: flake` contains lots of other flaky issues that we may not be interested in.

File: app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+  static const String _modifyType = 'blob';
+
+  static const int _kGracePeriodForClosedFlake = 15; // days
+
+  @override
+  Future<Body> get() async {
+    try {
+      final GitHub client = await config.createGitHubClientWithToken(await config.githubFlakyBotOAuthToken);
+      final Map<String, BuilderStats> nameToStats = await _getBuilderStats(client);
+      final Map<String, ExistingGithubIssue> nameToExistingIssue = await _getExistingGithubIssues(client);
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart:45]: nit: `nameTo ExistingIssues`

File: app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+      final Map<String, ExistingGithubPR> nameToExistingPR = await _getExistingGithubPRs(client);
+      final Set<String> importantFlakes = _getImportantFlakes(nameToStats.values.toList(), _threshold);
+      // Makes sure every the important flake has an github issue and a pr to
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart]: Nit: doc re what is an important flake.

File: app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+  Future<void> _maybeCloseIssue(Issue issue, GitHub client) async {
+    if (issue.assignee != null) {
+      return;
+    }
+    await client.issues.createComment(config.flutterSlug, issue.number, kCloseIssueComment);
+    await client.issues.edit(
+      config.flutterSlug,
+      issue.number,
+      IssueRequest(
+        state: 'closed',
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart:240]: Does it make sense if we leave the comment to suggest a fix is validated, and create a PR to mark the test as unflaky?
As for closing the issue, it maybe good to leave the decision to the bug owner?

File: app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+    // _kGracePeriodForClosedFlake days. It takes time for the flaky ratio to go
+    // down after the fixed is merged.
+    Issue issue;
+    if (!existingIssues.containsKey(newStats.name) ||
+        (existingIssues[newStats.name].issue.state == 'closed' &&
+         existingIssues[newStats.name].issue.closedAt.difference(DateTime.now())> const Duration(days: _kGracePeriodForClosedFlake))) {
+      if (!isImportant) {
+        throw 'This handler should only create new issue for important flake, something went wrong!';
+      }
+      issue = await _fileNewIssue(
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart:207]: Also a PR to mark the test as flaky?

File: app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github_utils.dart [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+      builder_name not like '%Beta%'
+group by builder_name;
+''';
+
+// Issue
+const String kCloseIssueComment = '''
+Closed this issue because the flaky rate of this test has dropped to 0.0% and is considered
+to be no longer flaky.
+
+If you think otherwise, reopen this issue and assign an owner to prevent it from being closed.
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github_utils.dart:37]: yes

File: app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart (hunk 1 above)
Comment (chunhtai) [reply app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart:381]: the line 394 is to deal with the github format. the actual line break in `contents.file.content` is encoded in base64, so it won't be replace there

File: app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart (hunk 7 above)
Comment (chunhtai) [reply app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart:207]: yes that is in line 222

**Pull request closed without merging**